### PR TITLE
Fix History and Bookmarks and ShieldsPanel and 3-Dot menu URL-Eliding

### PR DESCRIPTION
**Pull request by Brandon-T**
Resolves https://github.com/brave/internal/issues/1217

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: ios/brave-ios/Sources/Shared/Extensions/URLFormatterExtension.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import BraveCore
+import SwiftUI
+import UIKit
+
+extension URLFormatter {
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/Shared/Extensions/URLFormatterExtension.swift:10]: Please put this in `BraveShared` not `Shared`

File: ios/brave-ios/Sources/Shared/Extensions/URLFormatterExtension.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    let paragraphStyle = NSMutableParagraphStyle()
+    paragraphStyle.lineBreakMode = lineBreakMode
+    paragraphStyle.baseWritingDirection = .leftToRight
+
+    attributedString.setAttributes(
+      .init([
+        .font: font,
+        .paragraphStyle: paragraphStyle,
+      ])
+    )
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/Shared/Extensions/URLFormatterExtension.swift:27]: There's no need to create an attributed string and apply attributes after the fact, just create the string with the attributes immediately from the init

File: ios/brave-ios/Sources/Shared/Extensions/URLFormatterExtension.swift [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import BraveCore
+import SwiftUI
+import UIKit
+
+extension URLFormatter {
+  public static func createAttributedString(
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/Shared/Extensions/URLFormatterExtension.swift:11]: I think you're better off creating a custom `Text` view for this instead btw. something like:

```swift
struct URLElidedText: View {
  var content: String

  @Environment(\.font) private var font: Font

  private var paragraphStyle: NSParagraphStyle {
    ...
  }

  var body: some View {
    Text(AttributedString("\u{200E}\(string)", attributes: [.font: font, paragraphStyle: paragraphStyle]))
  }
}```

Comment (kylehickinson) [reply ios/brave-ios/Sources/Shared/Extensions/URLFormatterExtension.swift]: Didn't address this ☝️

Comment (Brandon-T) [reply ios/brave-ios/Sources/Shared/Extensions/URLFormatterExtension.swift]: I only saw this after. Just pushed a commit to address it.

File: ios/brave-ios/Sources/BraveUI/SwiftUI/URLElidedTextView.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import BraveCore
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/BraveUI/SwiftUI/URLElidedTextView.swift:6]: `BraveUI` doesn't have a dependency on `BraveCore` and probably shouldn't, maybe better to keep this in BraveShared despite it being UI

File: ios/brave-ios/Sources/BraveUI/SwiftUI/URLElidedTextView.swift [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  public var body: some View {
+    Text(
+      AttributedString(
+        "\u{200E}\(text)",  // LRM character prevents Text elements from rendering RTL special characters
+        attributes: .init([.font: font ?? .body, .paragraphStyle: paragraphStyle])
+      )
+    )
+    .font(font ?? .body)
+    .truncationMode(truncationMode)
```
Comment (kylehickinson) [inline ios/brave-ios/Sources/BraveUI/SwiftUI/URLElidedTextView.swift:40]: These are not needed, they're already part of the environment

**Review by kylehickinson — CHANGES_REQUESTED** [review]

**Review by stoletheminerals — APPROVED** [review]

**Review by kylehickinson — APPROVED** [review]

**Comment by kjozwiak** [comment]
Verification **PASSED** on `iPhone 11` running `iOS 18` using the following build(s):

```
Brave | 1.73.52 Chromium: 131.0.6778.14 (Official Build) nightly (64-bit) 
--- | ---
Revision | 6996738c6195...
OS | iOS
```

Verification notes can be found via https://github.com/brave/internal/issues/1217#issuecomment-2443047245.

**Pull request merged**